### ISSUE TITLE
Restic prune fails with "index is not complete"

### ISSUE THREAD

**Issue by gboor** [restic/restic#3551]
Output of `restic version`
--------------------------
restic 0.12.0 compiled with go1.15.8 on linux/amd64

How did you run restic exactly?
-------------------------------
I deleted some local files I also wanted removed from the remote backup (on backblaze), then I ran a new backup and afterwards a `forget --keep-last 1 --prune`, it was there that the first error popped up;

```
[0:00] 100.00%  4 / 4 files deleted
4 snapshots have been removed, running prune
loading indexes...
loading all snapshots...
finding data that is still in use for 1 snapshots
[0:00] 0.00%  0 / 1 snapshots
blob 44f81b811e92b584b7c997f2fa76decefb4d0b4c82542849ded564641b42b50b returned invalid hash
github.com/restic/restic/internal/repository.(*Repository).LoadBlob
        /restic/internal/repository/repository.go:210
github.com/restic/restic/internal/repository.(*Repository).LoadTree
        /restic/internal/repository/repository.go:728
github.com/restic/restic/internal/restic.loadTreeWorker
        /restic/internal/restic/tree_stream.go:37
github.com/restic/restic/internal/restic.StreamTrees.func1
        /restic/internal/restic/tree_stream.go:164
golang.org/x/sync/errgroup.(*Group).Go.func1
        /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20200625203802-6e8e738ad208/errgroup/errgroup.go:57
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1374
```

Now when I just run the `prune` command I get;

```
loading indexes...
loading all snapshots...
finding data that is still in use for 1 snapshots
[0:06] 100.00%  1 / 1 snapshots
searching used packs...
{<data/06a953fe>} not found in the index

Integrity check failed: Data seems to be missing.
Will not start prune to prevent (additional) data loss!
Please report this error (along with the output of the 'prune' run) at
https://github.com/restic/restic/issues/new/choose
Fatal: index is not complete
```

When delete the local cache and run an `ls`, all files appear to be there, but it looks like there is some corruption somewhere.

I also tried `rebuild-index`, which completed fine, but deleted over 3000 files. After that I run into issues running `ls` as well;

```
decrypting blob c20912425dfa4c7578740405df1c91d130641c57548d7e2e97110e9c5677994e failed: ciphertext verification failed
github.com/restic/restic/internal/repository.(*Repository).LoadBlob
        /restic/internal/repository/repository.go:204
github.com/restic/restic/internal/repository.(*Repository).LoadTree
        /restic/internal/repository/repository.go:728
github.com/restic/restic/internal/walker.walk
        /restic/internal/walker/walker.go:102
github.com/restic/restic/internal/walker.walk
        /restic/internal/walker/walker.go:122
github.com/restic/restic/internal/walker.walk
        /restic/internal/walker/walker.go:122
github.com/restic/restic/internal/walker.walk
        /restic/internal/walker/walker.go:122
github.com/restic/restic/internal/walker.Walk
        /restic/internal/walker/walker.go:51
main.runLs
        /restic/cmd/restic/cmd_ls.go:203
main.glob..func15
        /restic/cmd/restic/cmd_ls.go:44
github.com/spf13/cobra.(*Command).execute
        /home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826
github.com/spf13/cobra.(*Command).ExecuteC
        /home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914
github.com/spf13/cobra.(*Command).Execute
        /home/build/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
main.main
        /restic/cmd/restic/main.go:98
runtime.main
        /usr/local/go/src/runtime/proc.go:204
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1374
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Backblaze B2

Expected behavior
-----------------
No errors

Actual behavior
---------------
Errors

Steps to reproduce the behavior
-------------------------------
Unknown - I assume this should just work. I would've put this on the forum if the error itself did not specifically state to post it here.

Do you have any idea what may have caused this?
-----------------------------------------------
Remote-side corruption? Hard to imagine, though.

Do you have an idea how to solve the issue?
-------------------------------------------
I still have all the local data, so I could redo the entire backup. But since it's 7TB and took 2 weeks last time, I would much prefer some other workaround where the data already on the remote storage is reused.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Restic always makes me happy :)

**Comment by MichaelEischer** [comment]
Please run `restic check` to get more details on the repository damage. You might also want to update to restic 0.12.1.

Please lookup the packfile containing the tree blob using `restic find --show-pack-id --tree c20912425dfa4c7578740405df1c91d130641c57548d7e2e97110e9c5677994e`. Then use `restic cat --quiet pack <pack-id> | shasum -a 256` to verify whether the pack is intact.

> ```
> blob 44f81b811e92b584b7c997f2fa76decefb4d0b4c82542849ded564641b42b50b returned invalid hash
> ```

That error indicates the problems are quite likely caused by the backup client and not the remote storage. As restic only uploads cryptographically verified data, which is verified before checking the hash of a blob, this makes it essentially impossible to cause that error by data corruption at the remote storage. Instead the error has to be introduced at the client-side while assembling the backup data. This is usually caused by bit-flips in hardware. Please run some memory and cpu stress test like prime95 and memtest86.

Which Linux kernel version are you using? We've seen some problems in the past with older versions of Linux 5.2-5.4.

**Comment by gboor** [comment]
Thanks for the follow-up. I upgraded to restic 0.12.1 and that seems to have solved all the issues I previously had...

Just to be safe, I did run the check and the shasum and they're both fine. I ran another backup just to make sure I wasn't missing any files and that also went fine.

I finally ran the prune again and this time it completed without any errors during the rebuilding phase.

It should be noted that I deleted the cache before doing any of these things. I did so before as well, but now everything works. Apparently whatever the issue was is solved by the new version of Restic.

I will run prime95 and memtest as well, just to be sure, but it looks like my issue was fixed just by upgrading. Thanks for checking it out, though!
